build.cake

```diff
@@ -0,0 +1,497 @@
+	.Does(() =>
+{
+	if(FileExists("paket.exe"))
+		StartProcess("paket.exe", "restore");
+	else
+		StartProcess(@".paket\paket.exe", "restore");
+});
+
+Task("Version").Description("Get the version using GitVersion")
+	.IsDependentOn("Clean")
+	.Does(() =>
+{
+	var version = GitVersion();
+	_semanticVersionShort = version.Major + "." + version.Minor + "." + version.CommitsSinceVersionSource;
+	_semanticVersionLong = _semanticVersionShort + "+" + version.Sha + "." + version.BranchName;
+	Information("Semantic version (short):" + _semanticVersionShort);
+	Information("Semantic version (long ):" + _semanticVersionLong);
+
+	//Update all the variables now that we know the version number
+	var postFix = (version.BranchName == "master") ? string.Empty : "-" + version.Sha.Substring(0,8) + "." + NormalizeBrancheName(version.BranchName);
+	_zipFilename = string.Format(ZipFileTemplate, _semanticVersionShort + postFix);
+	_zipFilePath = System.IO.Path.Combine(buildAssetPath, _zipFilename);
+	_downloadUrl = string.Format(DownloadUrlTemplate, _semanticVersionShort) + _zipFilename;
+	_releaseVersion = "v" + _semanticVersionShort;
+});
+
+string NormalizeBrancheName(string branchName)
+{
+	return branchName.Replace('/', '_').Replace('\\', '_');
+}
+
+Task("UpdateAssemblyInfo").Description("Update AssemblyInfo properties with the Git Version")
+	.IsDependentOn("Version")
+	.Does(() =>
+{
+	if(BuildSystem.IsRunningOnAppVeyor)
+	{
+		Information("Updating Appveyor version to... " + _semanticVersionShort);
+		AppVeyor.UpdateBuildVersion(_semanticVersionShort);
+	}
+	
+	CreateAssemblyInfo("CommonAssemblyInfo.cs", new AssemblyInfoSettings {
+		Company="SEP",
+		Product = "GitTfs",
+		Copyright = "Copyright © 2009-" + DateTime.Now.Year,
+		Version = _semanticVersionShort,
+		FileVersion = _semanticVersionShort,
+		InformationalVersion = _semanticVersionLong
+	});
+});
+
+Task("Build").Description("Build git-tfs")
+	.IsDependentOn("Restore-NuGet-Packages")
+	.IsDependentOn("UpdateAssemblyInfo")
+	.Does(() =>
+{
+	// Use MSBuild
+	// /logger:"C:\Program Files\AppVeyor\BuildAgent\Appveyor.MSBuildLogger.dll" /nologo /p:BuildInParallel=true /m:4
+	MSBuild(PathToSln, settings => {
+
+		settings.SetConfiguration(Configuration)
+			.SetVerbosity(Verbosity.Minimal)
+			.SetMaxCpuCount(4);
+		if(_buildAllVersion)
+		{
+			settings.WithTarget("GitTfs_Vs2010")
+				.WithTarget("GitTfs_Vs2012")
+				.WithTarget("GitTfs_Vs2013");
+		}
+		settings.WithTarget("GitTfs_Vs2015")
+			.WithTarget("GitTfsTest");
+	});
+});
+
+void SetGitUserConfig()
+{
+	if(BuildSystem.IsRunningOnAppVeyor)
+	{
+		Information("Setting git user config to run some integration tests...");
+		//Merge with libgit2sharp now require having user name and email to be set!
+		StartProcess("git.exe", "config --global user.name \"git-tfs user for merge in unit tests\"");
+		StartProcess("git.exe", "config --global user.email \"git-tfs@unit-tests.com\"");
+	}
+}
+
+Task("Run-Unit-Tests").Description("Run the unit tests")
```

Comment:
This task might benefit from using a debug build, too, since IIRC the debug build will make it so that stack traces include line numbers, which I find helpful when debugging a test failure.